crates/bevy_ui/src/flex/mod.rs

```diff
@@ -117,45 +117,76 @@ without UI components as a child of an entity with UI components, results may be
 
     pub fn update_window(&mut self, window: &Window) {
         let stretch = &mut self.stretch;
-        let node = self.window_nodes.entry(window.id()).or_insert_with(|| {
-            stretch
-                .new_node(stretch::style::Style::default(), Vec::new())
-                .unwrap()
-        });
-
-        stretch
-            .set_style(
-                *node,
-                stretch::style::Style {
-                    size: stretch::geometry::Size {
-                        width: stretch::style::Dimension::Points(window.physical_width() as f32),
-                        height: stretch::style::Dimension::Points(window.physical_height() as f32),
-                    },
-                    ..Default::default()
-                },
-            )
-            .unwrap();
+        if let Some(window_nodes) = self.window_nodes.get(&window.id()) {
+            for (window_node, _) in window_nodes.iter() {
+                stretch
+                    .set_style(
+                        *window_node,
+                        stretch::style::Style {
+                            size: stretch::geometry::Size {
+                                width: stretch::style::Dimension::Points(
+                                    window.physical_width() as f32
+                                ),
+                                height: stretch::style::Dimension::Points(
+                                    window.physical_height() as f32
+                                ),
+                            },
+                            ..Default::default()
+                        },
+                    )
+                    .unwrap();
+            }
+        }
     }
 
     pub fn set_window_children(
         &mut self,
         window_id: WindowId,
         children: impl Iterator<Item = Entity>,
     ) {
-        let stretch_node = self.window_nodes.get(&window_id).unwrap();
-        let child_nodes = children
+        let children = children
             .map(|e| *self.entity_to_stretch.get(&e).unwrap())
             .collect::<Vec<stretch::node::Node>>();
-        self.stretch
-            .set_children(*stretch_node, child_nodes)
-            .unwrap();
+        let window_nodes = self.window_nodes.entry(window_id).or_default();
+
+        // Remove the previous nodes of deleted entities
+        let mut i = 0;
+        while i != window_nodes.len() {
+            if !children.contains(&window_nodes[i].1) {
+                self.stretch.remove(window_nodes[i].0);
+                self.stretch.remove(window_nodes[i].1);
+                window_nodes.remove(i);
+            } else {
+                i += 1;
+            }
+        }
+
+        // Add the nodes for new entities
+        for child in children.iter() {
```

Comment:
Ideally we can somehow cut down on the amount of scanning / nested looping happening here. For large numbers of root nodes this might start registering on frame time measurements.

However I'm not quite sure we have better options right now. The alternative is doing a bunch of hashing every frame for each root node (plus a full scan to detect removals).

Maybe we can work something out with Bevy queries?

Reply:
I'm looking into changing `flex_node_system` to only query for the bare minimum. The second commit adds a UiRoot component that may make it easier.